web/src/engine/osk/src/input/gestures/browser/subkeyPopup.ts

```diff
@@ -143,34 +142,56 @@ export default class SubkeyPopup implements GestureHandler {
     ss.fontSize=computedStyle.fontSize;
     ss.visibility='hidden';
 
-    var nKeys=subKeySpec.length,nRows,nCols;
-    nRows=Math.min(Math.ceil(nKeys/9),2);
-    nCols=Math.ceil(nKeys/nRows);
-
-    this.menuWidth = (nCols*e.offsetWidth + nCols*SUBKEY_DEFAULT_MARGIN_LEFT);
-    ss.width = this.menuWidth+'px';
+    const nKeys = subKeySpec.length;
+    // If we have more than 9 keys we put them in (at least) two rows.
+    const nRows=Math.min(Math.ceil(nKeys/9),2);
+    const nCols=Math.ceil(nKeys/nRows);
```

Comment:
Why is this logic being preserved?  We _very clearly_ throw it out in cases like the "MANY_KEYS" user test.

I do see a reference to `nCols` later, but I can't help but feel like we should be able to remove the reference and clean things up a bit further, given that "MANY_KEYS" is defined and seems to handle well.

Thinking about it... is this basically being used as a way to say "don't go over 9-wide" as a default, with the later handling an override for earlier "new row" construction when subkeys are wider?

Reply:
As I understand it, the previous code put more than 9 keys in two rows, I assume to reduce travel distance. This change preserves this logic even when we have a very wide osk where all keys would fit in one row. 

I could change it to:

```suggestion
    // Put a maximum of 9 keys in a row to reduce travel distance
    const nRows=Math.ceil(nKeys/9);
    const nCols=Math.ceil(nKeys/nRows);
```

This would prevent rows from having more than 9 keys even if they would fit. So this would be cleaner. However, with the previous code we would have a maximum of two rows - which in a way smells, but on the other hand we don't check that the number of rows actually fit and doesn't get cut off at the top.

I assume the previous implementation was done that way because we always have at least space for two rows when we display the submenu (the current row and some space above, either for the input field or for the banner).

In any case, having more than 18 keys in a longpress menu that would fit in two rows is probably rare enough that it doesn't matter too much which way we implement it.

@jahorton Which way do you prefer? This suggestion, or the way it's currently in the PR?

Reply:
> I assume the previous implementation was done that way because we always have at least space for two rows when we display the submenu (the current row and some space above, either for the input field or for the banner).

Yes, I think that would have been the case, though we should have documented it!

Reply:
@ermshiperete 

I'd prefer... neither.  We should be able to outright erase `nRows` and `nCols` based on the rest of your implementation.  I suppose we'd need a small tweak to prevent the column count from going _over_ 9 standard key-widths wide, but... that's the only change I think you'd need.

Refer to this line:

```
if (thisRowWidth + subkeyWidth + SUBKEY_DEFAULT_MARGIN_LEFT > vkbd.width) {
```

Replace `vkbd.width` with `Math.min(vkbd.width, e.offsetWidth * 9 + 8 * SUBKEY_DEFAULT_MARGIN_LEFT)`.  This would allow you to drop lines 155 -159, which have the final reference to `nCols`:

```
if (iCol >= nCols) {
  iRow++;
  iCol = 0;
  thisRowWidth = 0;
}
```

The second parameter from my suggestion above matches the width of 9 standard subkeys:  9 keys + 8 padding sections between them.  With this, we don't need something to force a new row after we reach number 9... unless someone's made narrow subkeys, I guess.  But... in that case, more could fit anyway, so they'll still be visible.  I don't _think_ we have many released keyboards with touch-layouts that have specialized subkey widths, so I'm not too worried about it.

As `nCols` itself is the final reference to `nRows`, this change allows dropping both.

Reply:
As for the _code comment_, I do prefer the suggested comment change.  I'd move it to precede whatever establishes parameter 2's value - the `e.offsetWidth * 9 + 8 * SUBKEY_DEFAULT_MARGIN_LEFT` bit - as that's what it'd best match in my prior comment's suggestion.

Reply:
Yes, that would work. However, we loose the balancing of the rows. Previously, if we had 10 keys we would display them in two rows with 5 keys each. The suggested algorithm would put them in two rows as well, but the first row would have 9 keys (if they fit), the second only one key.

So I think I leave the current implementation. I noticed a small flaw regarding the margins, so I fixed that (together with a small refactoring).